library/ssl_tls13_generic.c

```diff
@@ -4640,58 +4640,129 @@ void mbedtls_ssl_conf_early_data( mbedtls_ssl_config *conf, int early_data, char
+
+
+/*
+ * Implementation
+ */
+
+int mbedtls_ssl_new_session_ticket_process( mbedtls_ssl_context* ssl )
+{
+    int ret;
+    unsigned char* buf = NULL;
+    size_t buflen = 0;
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse new session ticket" ) );
+
+    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_new_session_ticket_fetch( ssl, &buf, &buflen ) );
+
+    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_new_session_ticket_parse( ssl, buf, buflen ) );
+
+    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_new_session_ticket_postprocess( ssl, ret ) );
+
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+    /* TBD: Return ACK message */
+    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+        mbedtls_ssl_recv_flight_completed( ssl );
+#endif /* MBEDTLS_SSL_PROTO_DTLS */
+
+cleanup:
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= parse new session ticket" ) );
+    return( ret );
+}
+
+static int mbedtls_ssl_new_session_ticket_fetch( mbedtls_ssl_context* ssl,
+                                  unsigned char** dst,
+                                  size_t* dstlen )
+{
+    *dst = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl ); 
+    *dstlen = ssl->in_msglen - mbedtls_ssl_hs_hdr_len( ssl );
+    
+    return( 0 );
+}
+
+static int mbedtls_ssl_new_session_ticket_parse( mbedtls_ssl_context* ssl,
+                                  unsigned char* buf,
+                                  size_t buflen )
 {
     int ret;
-    uint32_t lifetime, ticket_age_add;
     uint8_t ticket_nonce_len;
-    size_t ticket_len, ext_len;
+    uint16_t ticket_len, ext_len;
     unsigned char *ticket;
-    const unsigned char *msg;
     const mbedtls_ssl_ciphersuite_t *suite_info;
-    unsigned int msg_len;
-
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse new session ticket" ) );
+    size_t used = 0;
 
-    msg = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl );
-    msg_len = ( ssl->in_msg[2] << 8 ) | ssl->in_msg[3];
+    /*
+     * struct {
+     *    uint32 ticket_lifetime;
+     *    uint32 ticket_age_add;
+     *    opaque ticket_nonce<0..255>;
+     *    opaque ticket<1..2^16-1>;
+     *    Extension extensions<0..2^16-2>;
+     * } NewSessionTicket;
+     *
+     */
+    used += 4   /* ticket_lifetime */
+          + 4   /* ticket_age_add */
+          + 1   /* ticket_nonce length */
+          + 2   /* ticket length */
+          + 2;  /* extension length */
 
-    if( msg_len+ mbedtls_ssl_hs_hdr_len( ssl ) != ssl->in_msglen )
+    if( used > buflen )
     {
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
-        return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
+         MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
+         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
     }
 
     /* Ticket lifetime */
-    lifetime = ( msg[0] << 24 ) | ( msg[1] << 16 ) |
-        ( msg[2] << 8 ) | ( msg[3] );
+    ssl->session->ticket_lifetime =
+        ( buf[0] << 24 ) | ( buf[1] << 16 ) |
```

Comment:
This requires an explicit cast to `unsigned` prior to the shifts. Otherwise, `buf[x]` gets promoted to a _signed_ integer by the C standard, for which the shift `<< 24` is potentially undefined behaviour.

Reply:
See below for the ticket age where it's correct.